File: 2026-03/2026-03-27/truncate_text_2.py
CHARACTER_WIDTHS = {
    'ilI': 1,
    'fjrt': 2,
    'abcdeghkmnopqrstuvwxyzJL': 3,
    'ABCDEFGHKMNOPQRSTUVWXYZ': 4,
    ' ': 2,
    '.': 1
}
MAX_WIDTH = 50
TRUNCATION_STRING = "..."

def truncate_text(s):
    widths_list = [find_character_width(character) for character in s]

    total_units = sum(widths_list)

    if total_units <= 50:
        return s

    truncation_string_width = sum([find_character_width(character) for character in TRUNCATION_STRING])

    while total_units > MAX_WIDTH - truncation_string_width:
        total_units -= widths_list[-1]
        widths_list = widths_list[:-1]

    return s[:len(widths_list)] + TRUNCATION_STRING

def find_character_width(character):
    return [width for characters, width in CHARACTER_WIDTHS.items() if character in characters][0]

# --- TEST SUITE ---

tests_text = r'''
1. truncate_text("The quick brown fox") should return "The quick brown f...".
2. truncate_text("The silky smooth sloth") should return "The silky smooth s...".
3. truncate_text("THE LOUD BRIGHT BIRD") should return "THE LOUD BRIG...".
4. truncate_text("The fast striped zebra") should return "The fast striped z...".
5. truncate_text("The big black bear") should return "The big black bear".
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

    print("----------------------------",
        f"\n⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "\n🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)

File: 03 March/03-27/truncateText_2.js
const CHARACTER_WIDTHS = {
  'ilI': 1,
  'fjrt': 2,
  'abcdeghkmnopqrstuvwxyzJL': 3,
  'ABCDEFGHKMNOPQRSTUVWXYZ': 4,
  ' ': 2,
  '.': 1
};

function truncateText(str) {

  const MAX_WIDTH = 50;
  const TRUNCATION_STRING = "...";

  let widthsArray = str
    .split("")
    .map(character => findCharacterWidth(character));

  let totalUnits = widthsArray.reduce((total, width) => total + width);

  if (totalUnits <= 50)
    return str;

  const truncationStringWidth = TRUNCATION_STRING
    .split("")
    .reduce((total, character) => total + findCharacterWidth(character), 0);

  while (totalUnits > MAX_WIDTH - truncationStringWidth) {
    totalUnits -= widthsArray.at(-1);
    widthsArray = widthsArray.slice(0, -1);
  }

  return str.slice(0, widthsArray.length) + TRUNCATION_STRING;
}

const findCharacterWidth = character => {
  for (const [characters, width] of Object.entries(CHARACTER_WIDTHS)) {
    if (characters.includes(character))
      return width;
  }
};

// --- TEST SUITE ---

const testsText = String.raw`
1. truncateText("The quick brown fox") should return "The quick brown f...".
2. truncateText("The silky smooth sloth") should return "The silky smooth s...".
3. truncateText("THE LOUD BRIGHT BIRD") should return "THE LOUD BRIG...".
4. truncateText("The fast striped zebra") should return "The fast striped z...".
5. truncateText("The big black bear") should return "The big black bear".
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);

    const comparison = Array.isArray(testOutput)
      ? arraysEqual(functionCallOutput, testOutput)
      : functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      ++failCount;
    }
  })

  console.log("----------------------------",
    failCount
    ? `\n⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "\n🎉SUCCESS: All tests PASSED."
  );
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (let i = 0; i < a.length; ++i) {
    if (Array.isArray(a[i])) {
      if (Array.isArray(b[i])) {
        if (!arraysEqual(a[i], b[i]))
          return false;
      }
      else 
        return false;
    }

    else if (a[i] !== b[i])
      return false;
  }

  return true;
}

runTests(testData);